File: examples/yolop/cpp/yolop.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <iostream>
#include <memory>
#include <string>

#include <opencv2/opencv.hpp>

#include "vision_service.h"

namespace {

void usage(const char* program) {
    std::cout
        << "Usage: " << program << " <config.yaml> [options]\n"
        << "  --model-path <path>  Override model_path\n"
        << "  --image <path>       Override test_image\n"
        << "  --output <path>      Output image (default: yolop_result.jpg)\n";
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        usage(argv[0]);
        return 1;
    }
    const std::string config_path = argv[1];
    std::string model_path;
    std::string image_path;
    std::string output_path = "yolop_result.jpg";
    for (int i = 2; i < argc; ++i) {
        const std::string argument = argv[i];
        if (argument == "--model-path" && i + 1 < argc) {
            model_path = argv[++i];
        } else if (argument == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (argument == "--output" && i + 1 < argc) {
            output_path = argv[++i];
        } else if (argument == "--help") {
            usage(argv[0]);
            return 0;
        } else {
            std::cerr << "Unknown or incomplete argument: " << argument << std::endl;
            return 1;
        }
    }

    std::unique_ptr<VisionService> service =
        VisionService::Create(config_path, model_path, false);
    if (!service) {
        std::cerr << "Create failed: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }
    if (image_path.empty()) image_path = service->GetDefaultImage();
    cv::Mat image = cv::imread(image_path);
    if (image.empty()) {
        std::cerr << "Could not read image: " << image_path << std::endl;
        return 1;
    }

    VisionServiceResponse response;
    const VisionServiceStatus status = service->Infer(image, &response);
    if (status != VISION_SERVICE_OK || !response.ok) {
        std::cerr << "Infer failed: " << service->LastError() << std::endl;
        return 1;
    }
    size_t detections = 0;
    size_t masks = 0;
    for (const vision::Result& result : response.results) {
        detections += std::holds_alternative<vision::Detection>(result) ? 1 : 0;
        masks += std::holds_alternative<vision::Segmentation>(result) ? 1 : 0;
    }
    if (masks != 2) {
        std::cerr << "Expected two semantic masks, got " << masks << std::endl;
        return 1;
    }
    std::cout << "Detections: " << detections << ", masks: " << masks << std::endl;
    if (!service->SupportsDraw()) {
        std::cerr << "YOLOP must support Draw" << std::endl;
        return 1;
    }
    cv::Mat output;
    if (service->Draw(image, response, &output) != VISION_SERVICE_OK || output.empty()) {
        std::cerr << "Draw failed: " << service->LastError() << std::endl;
        return 1;
    }
    if (!cv::imwrite(output_path, output)) {
        std::cerr << "Could not write output: " << output_path << std::endl;
        return 1;
    }
    std::cout << "Saved: " << output_path << std::endl;
    return 0;
}

File: examples/yolop/python/yolop.py
#!/usr/bin/env python3
# Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
# SPDX-License-Identifier: Apache-2.0

"""YOLOP multi-task inference through the unified VisionService binding."""

import argparse
from pathlib import Path

import cv2

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def parse_args():
    parser = argparse.ArgumentParser(description="YOLOP example")
    parser.add_argument(
        "--config",
        default=str(Path(__file__).parent.parent / "config" / "yolop.yaml"),
    )
    parser.add_argument("--model-path", default="")
    parser.add_argument("--image")
    parser.add_argument("--output", default="yolop_result.jpg")
    parser.add_argument("--conf-threshold", type=float, default=-1.0)
    parser.add_argument("--iou-threshold", type=float, default=-1.0)
    return parser.parse_args()


def main():
    args = parse_args()
    service = VisionServiceNative.create(
        args.config, model_path_override=args.model_path
    )
    try:
        image_path = args.image or service.get_default_image()
        if not image_path:
            raise ValueError("No image; use --image or set test_image in config")
        image = cv2.imread(str(Path(image_path).expanduser()))
        if image is None:
            raise FileNotFoundError(f"Could not read image: {image_path}")
        status, results = service.infer_image(
            image, conf=args.conf_threshold, iou=args.iou_threshold
        )
        if status != VisionServiceStatus.OK:
            raise RuntimeError(service.last_error())
        detections = [result for result in results if result.mask is None]
        masks = [result for result in results if result.mask is not None]
        if len(masks) != 2:
            raise RuntimeError(f"Expected two semantic masks, got {len(masks)}")
        for result in masks:
            if result.mask.shape != image.shape[:2]:
                raise RuntimeError(
                    f"Unexpected mask shape {result.mask.shape}, expected {image.shape[:2]}"
                )
        if not service.supports_draw():
            raise RuntimeError("YOLOP must support draw")
        draw_status, output = service.draw(image)
        if draw_status != VisionServiceStatus.OK or output is None:
            raise RuntimeError(service.last_error())
        if not cv2.imwrite(args.output, output):
            raise RuntimeError(f"Could not write output: {args.output}")
        print(f"Detections: {len(detections)}, masks: {len(masks)}")
        print(f"Saved: {args.output}")
        return 0
    finally:
        service.release()


if __name__ == "__main__":
    raise SystemExit(main())
